### PR TITLE
Use CFFI to Reuse Code

### LINKED ISSUE #2710: Duplication on attempt_nocopy_reshape logic
With https://github.com/numba/numba/pull/2702, there will be two translation of the numpy's attempt_nocopy_reshape.  The other one is in [_helperlib.c](https://github.com/numba/numba/blob/b4e44a98dcc2b4f1036d58f5aab39542dc1ffb41/numba/_helperlib.c#L442).

### PR DESCRIPTION
**Pull request by njwhite**
Fix  #2710, if it works...

This PR could be made tidier with `numpy.ctypeslib.ndpointer` too.

### CODE REVIEW

File: numba/_helpermod.c [hunk 1]
```diff
@@ -85,6 +85,7 @@ build_c_helpers_dict(void)
     declmethod(get_pyobject_private_data);
     declmethod(set_pyobject_private_data);
     declmethod(reset_pyobject_private_data);
+    declmethod(attempt_nocopy_reshape);
```
Comment (sklam) [inline numba/_helpermod.c]: duplicated.  already in line 84.

File: numba/dummyarray.py [hunk 2]
```diff
@@ -4,13 +4,28 @@
...
 
+from . import _helperlib
+
 
 Extent = namedtuple("Extent", ["begin", "end"])
 
 
+attempt_nocopy_reshape = ctypes.CFUNCTYPE(
+    ctypes.c_int,
+    ctypes.c_long, # nd
```
Comment (sklam) [inline numba/dummyarray.py]: It's better to match the `npy_intp` types  with the `numpy.ctypeslib.c_intp`

**Comment by njwhite** [comment]
@sklam ah, that makes the logic much clearer!

**Comment by sklam** [comment]
Thanks for your patch!

**Pull request merged**